### PR TITLE
Refactor how extrapolation is performed

### PR DESCRIPTION
**Pull request by BenjaminRodenberg**
## Main changes of this PR

* Create Extrapolation class in `cplscheme/impl` for functionality.
* Move all current functionality of Waveform into Extrapolation. Waveform is reserved for interpolation. See #1029 
* Store Extrapolation in CouplingData.

## Motivation and additional information

We want to implement extrapolation and interpolation as described in #1029. This PR adds some changes from #1029 directly to develop and simplifies #1029.

## Author's checklist

* [ ] I added a changelog file with this PR number in `docs/changelog/` if there are noteworthy changes.
* [ ] I ran `tools/formatting/format-all` to ensure everything is formatted correctly.
* [ ] I sticked to C++14 features.
* [ ] I sticked to CMake version 3.10.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 1]
```diff
@@ -257,12 +253,12 @@ void BaseCouplingScheme::advance()
...
 }
 
-void BaseCouplingScheme::storeDataInWaveforms()
+void BaseCouplingScheme::storeExtrapolationData()
 {
   PRECICE_TRACE(_timeWindows);
   for (DataMap::value_type &pair : _allData) {
     PRECICE_DEBUG("Store data: {}", pair.first);
-    _waveforms[pair.first]->store(pair.second->values());
+    pair.second->storeDataInExtrapolation();
```
Comment (uekerman) [inline src/cplscheme/BaseCouplingScheme.cpp:261]:
```suggestion
    pair.second->storeExtrapolationData();
```
?
If we really use two different names for the two different functions, the name should explain the difference.

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 2]
```diff
@@ -450,15 +445,16 @@ void BaseCouplingScheme::checkCompletenessRequiredActions()
   }
 }
 
-void BaseCouplingScheme::setupDataMatrices()
+void BaseCouplingScheme::initializeStorage()
```
Comment (uekerman) [inline src/cplscheme/BaseCouplingScheme.cpp:448]:
```suggestion
void BaseCouplingScheme::initializeStorages()
```
?

File: src/cplscheme/CouplingData.hpp [hunk 3]
```diff
@@ -45,14 +47,24 @@ class CouplingData {
...
 private:
   /**
    * @brief Default constructor, not to be used!
    *
    * Necessary when compiler creates template code for std::map::operator[].
    */
   CouplingData()
-      : requiresInitialization(false)
+      : requiresInitialization(false),
+        _extrapolation(0)
```
Comment (uekerman) [inline src/cplscheme/CouplingData.hpp:67]: Wouldn't it be better to use `UNDEFINED_EXTRAPOLATION_ORDER` here as well?

File: src/cplscheme/impl/Extrapolation.cpp [hunk 4]
```diff
@@ -1,70 +1,73 @@
-#include "time/Waveform.hpp"
+#include "Extrapolation.hpp"
 #include <algorithm>
+#include "cplscheme/CouplingScheme.hpp"
 #include "logging/LogMacros.hpp"
 #include "utils/EigenHelperFunctions.hpp"
 
 namespace precice {
-namespace time {
+namespace cplscheme {
```
Comment (uekerman) [inline src/cplscheme/impl/Extrapolation.cpp:8]: IIRC we normally also use an `impl` namespace here.

**Review by uekerman — APPROVED** [review]
Looks good. Only minor things below. Feel free to directly merge if fixed and no more questions.

File: src/cplscheme/CouplingData.hpp (hunk 3 above)
Comment (BenjaminRodenberg) [reply src/cplscheme/CouplingData.hpp:67]: Yes. I will implement this :+1:

File: src/cplscheme/impl/Extrapolation.cpp (hunk 4 above)
Comment (BenjaminRodenberg) [reply src/cplscheme/impl/Extrapolation.cpp:8]: Yes. I will add this namespace.

File: src/cplscheme/BaseCouplingScheme.cpp (hunk 2 above)
Comment (BenjaminRodenberg) [reply src/cplscheme/BaseCouplingScheme.cpp]: Ok.

**Pull request merged**